mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala

```diff
@@ -324,13 +324,22 @@ class ALSModel private[ml] (
     // create a new column named map(predictionCol) by running the predict UDF.
     val validatedUsers = checkIntegers(dataset, $(userCol))
     val validatedItems = checkIntegers(dataset, $(itemCol))
+
+    val validatedInputAlias = Identifiable.randomUID("__als_validated_input")
+    val itemFactorsAlias = Identifiable.randomUID("__als_item_factors")
+    val userFactorsAlias = Identifiable.randomUID("__als_user_factors")
+
     val predictions = dataset
-      .join(userFactors,
-        validatedUsers === userFactors("id"), "left")
-      .join(itemFactors,
-        validatedItems === itemFactors("id"), "left")
-      .select(dataset("*"),
-        predict(userFactors("features"), itemFactors("features")).as($(predictionCol)))
+      .withColumns(Map($(userCol) -> validatedUsers, $(itemCol) -> validatedItems))
```

Comment:
shall we use the `Seq` version of `withColumns`? So that the column order is deterministic.

Reply:
why do we need a `withColumns` now?

Reply:
It maybe not needed, I want to use `withColumns` to validate the columns first (while keep the column name) and then reference the validated column by `s"${validatedInputAlias}.${$(itemCol)}"`